common/preconditioner/isai_kernels.hpp.inc

```diff
@@ -0,0 +1,214 @@
+                subwarp_id};  // write to trisystem in row major
+
+        while (m_row_ptr < m_row_end && i_row_ptr < i_row_end &&
+               idx < end_trisystem) {
+            const auto sparsity_col = i_col_idxs[i_row_ptr];
+            const auto m_col = m_col_idxs[m_row_ptr];
+            if (sparsity_col == m_col) {
+                trisystem[idx] = m_values[m_row_ptr];
+                idx += SubwarpSize;
+                ++m_row_ptr;
+                ++i_row_ptr;
+            } else if (m_col < sparsity_col) {
+                ++m_row_ptr;
+            } else {
+                ++i_row_ptr;
+                idx += SubwarpSize;
+            }
+        }
+    }
+
+    this_block.sync();
+
+    // Now, read a full col of `trisystem` into local registers, which will be
+    // row elements after this (implicit) transpose
+    ValueType local_row[SubwarpSize];
+    const auto trisystem_col_start = trisystem_start + subwarp_id * SubwarpSize;
+    for (int i = 0; i < SubwarpSize; ++i) {
+        const auto intermediate_idx = subwarp_id + i;
+        const auto coalesced_idx = intermediate_idx % SubwarpSize;
+        local_row[coalesced_idx] =
+            trisystem[trisystem_col_start + coalesced_idx];
+    }
+
+    const auto rhs = trs_solve(num_row_elems, local_row, subwarp);
+
+    // Write back:
+    if (subwarp_id < num_row_elems) {
+        const auto idx = i_row_begin + subwarp_id;
+        if (gko::isfinite(rhs)) {
+            i_values[idx] = rhs;
+        } else {
+            i_values[idx] =
+                i_col_idxs[idx] == row ? one<ValueType>() : zero<ValueType>();
+        }
+    }
+}
+
+
+template <int SubwarpSize, typename ValueType, typename IndexType>
+__global__ __launch_bounds__(default_block_size) void generate_l_inverse(
+    IndexType num_rows, const IndexType *__restrict__ m_row_ptrs,
+    const IndexType *__restrict__ m_col_idxs,
+    const ValueType *__restrict__ m_values, IndexType *__restrict__ i_row_ptrs,
+    IndexType *__restrict__ i_col_idxs, ValueType *__restrict__ i_values)
+{
+    auto trs_solve = [](IndexType num_elems,
+                        const ValueType *__restrict__ local_row,
+                        group::thread_block_tile<SubwarpSize> &subwarp) {
+        const int subwarp_id = subwarp.thread_rank();
+        ValueType rhs =
+            subwarp_id == num_elems - 1 ? one<ValueType>() : zero<ValueType>();
+        // Solve Triangular system
+        // Must be the full SubwarpSize, so shuffles work properly
+        for (int d_col = SubwarpSize - 1; d_col >= 0; --d_col) {
+            const auto elem = local_row[d_col];
+            if (d_col == subwarp_id) {
+                // make sure rhs is not NaN when subwarp is out of range
+                rhs = subwarp_id < num_elems ? rhs / elem : zero<ValueType>();
+            }
+
+            const ValueType bot = subwarp.shfl(rhs, d_col);
+            if (subwarp_id < d_col) {
+                rhs -= bot * elem;
+            }
+        }
```

Comment:
Just in general: Isn't this type of thing well-suited for `warp_blas.hpp.inc`?

Reply:
Since this requires a very specific setup, I wanted to keep it local in this file.